### PR TITLE
LRUD V3 - Major Rewrite

### PR DESCRIPTION
**Pull request by thomascgray**
_I cannot apologise enough for the size of this pull request. More than happy to sit and go through the changes with any interested party._

# Scenario

This is the V3 re-write of LRUD.

For further details, please see [the paper doc](https://paper.dropbox.com/doc/SSR-Controller-Module-LRUD-V3--AcSnSsWPAnJE1vDPNQJQEjSZAg-y0v9YqarEOXNP7R2151RK) but the general gist of what has changed;

- "real" tree structure
- cleaner/easier to understand "grid" functionality (now "index aligned" functionality)
- supporting a concept of "column span"/"column width"
- codified definition of which node is the “root” node
- column alignment between 2 of anything, not just sibling rows
- refactored code
- all leaf nodes have an index value, so we can understand the relative order with any context
- better handling/behaviour when unregistering a node

There is also some new stuff around the "infrastructure" of the project.

- LRUD is now written in Typescript and compiles down to ES5 via Typescript compilation & rollup.js
- modified `package.json` to point to some new things, etc.

There is also [a JIRA ticket](https://jira.dev.bbc.co.uk/browse/IPLAYERTVV1-8467) but it needs updating

# Changelog

- Typescript all the things!
- `index.js` is completely re-written to be a real class. Many of the concepts are the same, but almost entirely new function names, etc.
- tons of new functionality, based around everything described above
- all the tests have been re-written and broken apart into different files
- inclusion of some draw.io diagrams to illustrate expected behaviour for some particular test scenarios (the names match up with tags on the tests)
- changing from tiny-emitter to mitt, as it is more active, doesn't have any active vulnerability issues, and is smaller
- completely re-written readme and docs to cover lots of different behaviours and recipes for common setups
- new ADR

### CODE REVIEW

File: docs/index-align.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  .register('root', { orientation: 'vertical', isIndexAlign: true })
+  .register('row-1', { orientation: 'horizontal' })
+  .register('row-1_item-1', { isFocusable: true, parent: 'row-1' })
+  .register('row-1_item-2', { isFocusable: true, parent: 'row-1' })
+  .register('row-2', { orientation: 'horizontal' })
+  .register('row-2_item-1', { isFocusable: true, parent: 'row-2' })
+  .register('row-2_item-2', { isFocusable: true, parent: 'row-2' })
+```
+
+Now, when the user enters the 2nd row (whos active child, as its never been focused on, is its first item) they will instead focus on the node that has the same index as the node they just left.
```
Comment (jordanholt) [inline docs/index-align.md]: "whose" active child 😊

Comment (thomascgray) [reply docs/index-align.md]: You are ~write~ right - fixed.

File: docs/usage.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    .registerNode('Z', { parent: 'root', index: 2 })
+```
+
+### `isIndexAlign`
+
+`boolean`
+
+To be used in conjunction with orientation to give a node index alignment functionality.
+
+Index alignment ensures that when focus would move from one this nodes descendants to another, Lrud will attempt to ensure that the focused node matches the index of the node that was left.
```
Comment (jordanholt) [inline docs/usage.md]: Sorry, just little things... "one of this" here

File: src/index-align.test.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    navigation.handleKeyEvent({ direction: 'down' })
+    expect(navigation.currentFocusNodeId).toEqual('grid-b-row-1-col-2')
+
+    navigation.handleKeyEvent({ direction: 'down' })
+    expect(navigation.currentFocusNodeId).toEqual('grid-b-row-2-col-2')
+  })
+
+  test('with vertical index alignment', () => {
+
```
[the remark this answers is not in the document]
Comment (thomascgray) [reply src/index-align.test.js]: Good spot - will delete for now, and add more test scenarios as we get feedback from Moss Piglets

File: docs/usage.md (hunk 2 above)
Comment (thomascgray) [reply docs/usage.md]: No apologies needed - English are hard.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+    if (parentsChildPaths == null) {
+      const parentPath = this.getPathForNodeId(node.parent)
+      Set(this.tree, parentPath + '.activeChild', nodeId)
+    }
+
+    // if no `index` set, calculate it
+    if (!node.index) {
+      let parentNode = this.getNode(node.parent)
+      if (parentNode) {
+        let parentsChildren = this.getNode(node.parent).children
```
Comment (jordanholt) [inline src/index.ts]: Can `parentNode` and `parentsChildren` be `const` here?

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+   */
+  register(nodeId, node : any = {}) {
+    return this.registerNode(nodeId, node)
+  }
+
+  /**
+   * unregister a node from the navigation tree
+   * @param {string} nodeId
+   */
+  unregisterNode (nodeId) {
```
Comment (jordanholt) [inline src/index.ts]: Should unregistering a node also unregister any overrides that have this node as their target?

File: src/index.ts (hunk 4 above)
Comment (thomascgray) [reply src/index.ts]: Indeed they can 👍 fixed.

File: src/index.ts (hunk 5 above)
Comment (thomascgray) [reply src/index.ts]: Yep, I agree it should. Will wire up this behaviour and add some tests.

Comment (thomascgray) [reply src/index.ts]: Also added to this logic a little to unregister an `override` if its `target` OR `id` matches the ID of the node that was unregistered.

**Comment by MaxMilton** [comment]
Is that document you linked to strictly for BBC internal use? If not, I'm keen to read over it :)

**Comment by jordanholt** [comment]
@MaxMilton We'll have a review of it and see if we can get a copy to you.

**Comment by thomascgray** [comment]
@MaxMilton hey man - basically, everything that was in the internal paper doc has been formalised into the ADR that's included in this pull request. So feel free to take a look at that for our reasoning, etc. on the rebuild.

**Comment by MaxMilton** [comment]
Awesome. Thanks for your work on this, it's looking good :smile:

**Pull request merged**